File: src/models/task.py
"""Task SQLModel entity."""

from datetime import date, datetime

from sqlmodel import Field, SQLModel

VALID_PRIORITIES = ("high", "medium", "low")
VALID_CATEGORIES = ("work", "home", "personal", "health", "other")
VALID_RECURRENCES = ("daily", "weekly", "monthly", "yearly")
DEFAULT_REMINDER_MINUTES = 15


class Task(SQLModel, table=True):
    __tablename__ = "tasks"

    id: int | None = Field(default=None, primary_key=True)
    title: str = Field(min_length=1, max_length=500)
    description: str | None = Field(default=None)
    is_completed: bool = Field(default=False)
    priority: str = Field(default="medium")
    category: str | None = Field(default=None)
    due_date: date | None = Field(default=None)
    recurrence: str | None = Field(default=None)
    due_time: str | None = Field(default=None)
    reminder_minutes: int | None = Field(default=None)
    reminder_sent: bool = Field(default=False)
    created_at: datetime = Field(default_factory=datetime.now)
    updated_at: datetime = Field(default_factory=datetime.now)

File: frontend/src/types/task.ts
export type Priority = 'high' | 'medium' | 'low';

export type Category = 'work' | 'home' | 'personal' | 'health' | 'other';

export type Recurrence = 'daily' | 'weekly' | 'monthly' | 'yearly';

export type SortField = 'created_at' | 'due_date' | 'title' | 'priority';

export type SortOrder = 'asc' | 'desc';

export type TaskStatus = 'all' | 'active' | 'completed';

export interface Task {
  id: number;
  user_id: string;
  title: string;
  description?: string;
  is_completed: boolean;
  priority: Priority;
  category?: Category;
  due_date?: string;
  recurrence?: Recurrence;
  due_time?: string;
  reminder_minutes?: number;
  reminder_sent: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateTaskInput {
  title: string;
  description?: string;
  priority?: Priority;
  category?: Category;
  due_date?: string;
  due_time?: string;
  recurrence?: Recurrence;
  reminder_minutes?: number;
}

export type UpdateTaskInput = Partial<CreateTaskInput>;

export interface TaskFilters {
  status: TaskStatus;
  priority?: Priority;
  category?: Category;
  search?: string;
  sort_by: SortField;
  sort_order: SortOrder;
}

export interface TaskListResponse {
  tasks: Task[];
  total: number;
}

export interface ApiError {
  detail: string | Array<{
    loc: string[];
    msg: string;
    type: string;
  }>;
}